pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -1093,28 +1087,18 @@ func buildDefaultHTTPRoute(clusterName string, operation string) *route.Route {
 	return val
 }
 
-// setTimeout sets timeout for a route.
-func setTimeout(action *route.RouteAction, vsTimeout *duration.Duration, node *model.Proxy) {
+// SetRouteTimeout sets timeout for a route.
+func SetRouteTimeout(action *route.RouteAction, vsTimeout *duration.Duration) {
 	// Configure timeouts specified by Virtual Service if they are provided, otherwise set it to defaults.
 	action.Timeout = Notimeout
 	if vsTimeout != nil {
 		action.Timeout = vsTimeout
 	}
-	if node != nil && node.IsProxylessGrpc() {
-		// TODO(stevenctl) merge these paths; grpc's xDS impl will not read the deprecated value
-		action.MaxStreamDuration = &route.RouteAction_MaxStreamDuration{
-			MaxStreamDuration: action.Timeout,
-		}
-	} else {
+	action.MaxStreamDuration = &route.RouteAction_MaxStreamDuration{
+		MaxStreamDuration: action.Timeout,
 		// If not configured at all, the grpc-timeout header is not used and
 		// gRPC requests time out like any other requests using timeout or its default.
-		// Use deprecated value for now as the replacement MaxStreamDuration has some regressions.
-		// nolint: staticcheck
-		if action.Timeout.AsDuration().Nanoseconds() == 0 {
-			action.MaxGrpcTimeout = Notimeout
-		} else {
-			action.MaxGrpcTimeout = action.Timeout
-		}
+		GrpcTimeoutHeaderMax: action.Timeout,
```

Comment:
Yes, this current logic respects VS's timeout firstly, and if not set fallback to grpc-timeout header

Reply:
I think this is incorrect. The old configuration(`MaxGrpcTimeout` which is currently used by Istio) and the new replacement config `GrpcTimeoutHeaderMax`  both respect `grpc-timeout` header if present. So the `finalTimeout` that envoy uses is same as what is there in `grpc-timeout` header. The max value is only used to cap the timeout.

This is how the timeout is arrived for `GrpcTimeoutHeaderMax` https://github.com/envoyproxy/envoy/blob/a29f064197841454f4cc4de8a6c4a75e862d855d/source/common/http/conn_manager_impl.cc#L1306

and this is how old configuration `MaxGrpcTimeout` arrives at timeout  https://github.com/envoyproxy/envoy/blob/517d2dba98ba7691d1642caf2ef4dd42a6daf029/source/common/router/router.cc#L141

So the comment [here](https://github.com/istio/istio/issues/45141#issue-1726004599) is incorrect that it says, the new MaxStreamDuration flag, uses `0ms` as timeout. 

The only difference between new config and old config is - the new config returns the proper response code `DEADLINE_EXCEEDED` instead of `UNAVAILABLE` - that is what the new [implementation](https://github.com/envoyproxy/envoy/pull/13302) tried to fix.
But the new implementation has several other [issues](https://github.com/envoyproxy/envoy/issues/16129) . 

If we move to the new implementation as this PR is trying to do - we will loose the ability to set timeouts via Envoy filters as this sets up the new max stream duration to `0s` if there is no VS so it would completely disable timeouts.

So there is no better alternative here - both approaches have one or the other issue.

Reply:
The issuse claims the response status code.

> we will loose the ability to set timeouts via Envoy filters as this sets up the new max stream duration to 0s if there is no VS so it would completely disable timeouts.

VS is the first class API, it is encouraged to  use vs to set timeout. At least we have a way to do that.


I am not sure `MaxGrpcTimeout ` was set to 0 as well, why envoyfilter merge works for it

Reply:
> VS is the first class API, it is encouraged to use vs to set timeout. At least we have a way to do that.

In both the cases there is no VS involved

> I am not sure MaxGrpcTimeout  was set to 0 as well, why envoyfilter merge works for it

If MaxGrpcTimeout was set to 0, it wont interfere with regular time outs -it would only enable Envoy to use `grpc-timeout` header.  But if we set MaxStreamDuration to 0 that will disable timeouts on the route - the timeout field specified via Envoy filters is not used.

Reply:
I think it is still viable to mutate the new maxstreamDuration with envoyfilter

Reply:
We could but it breaks the existing Envoy filters. Please note we use that only for cases where notimeout is specified for other cases we use MaxGrpcTimeout header because MaxStreamDuration has those bugs. So in that case also we get the incorrect response code if clients set grpc-timeout header.  So the behaviour would be different if VS timeout is specified and grpc-timeout header is present vs. no VS timeout is specified (defaults to zero) and grpc-timeout header is present if we use maxStreamDuration of this case only. Ideally we want to use MaxStreamDuration in both so that we can send back the right header.

Reply:
I think it is better to fix this properly in Envoy so that we get the consistent behaviour. Any thing we do here has some inconsistency or other

Reply:
It is unacceptable that envoy deprecate a field with the unmature maxStreamDuration field. From enduser view, response code is very important.  Meanwhile we donot ever document envoyfilter is mature and compatable across all releases.

I am not sure how to fix both issues though, defer to you @ramaraochavali

Reply:
> It is unacceptable that envoy deprecate a field with the unmature maxStreamDuration field.

I agree but without that we can still use MaxGrpcTimeout that has same response code problem for cases where user sets VS timeout.

So it is not totally related to Envoy filters - even without envoy filters we have this inconsistency.

Let me look at the current gaps in Envoy and see if our cases are solved by some other means. I will update this thread.

Reply:
I think it is better to resolve it in envoy to get consistency